### PR TITLE
Guarantee that client leave and noClient messages are generated from deli

### PR DESCRIPTION
**Pull request by GaryWilber**
It is not guaranteed that deli will insert a leave & noClient message for a document in some specific cases. Here's one case:
1. The deli lambda is running and there's at least 1 client connected to the document.
2. Assume the deli checkpoint is fully up to date.
3. Let's say that the alfred processes crash. This means alfred did not insert a leave message for the clients that were connected

Deli normally handles this fine. It has an idle timer, which will cause it to detect that the client is no longer active and it will create a leave op for them. However that is not perfect.

4. If the deli lambda is stopped (due to a kafka rebalance or crash), the deli lambda will not start up again. It doesn't start up again because it already checkpointed it's latest kafka message

At this point we have a document in the system that won't have a leave op / noClient op inserted until a new client connects to it.

The fix is to make deli's Kafka checkpoint always stay behind by 1 message until it sends the noClient message. Once deli has no clients remaining and sends noClient, it will start making the kafka checkpoint the current latest message.
This makes it so deli will spawn again even after a rebalance.

Changes:
- Make deli's kafka checkpoint stay 1 behind until no clients are active
- Start the idle timer when deli is created
- Remove the queuedMessage from being included in the deli mongodb checkpoint
- Rehydrate `noActiveClients` property when deli is created. Before it was always set to `false` even though that may have not been correct.

### CODE REVIEW

File: server/routerlicious/packages/lambdas/src/deli/checkpointContext.ts [hunk 1]
```diff
@@ -74,7 +85,8 @@ export class CheckpointContext {
         if (checkpoint.clear) {
             updateP = this.checkpointManager.deleteCheckpoint(checkpoint);
         } else {
-            const deliCheckpoint: IDeliState = { ...checkpoint };
+            // clone the checkpoint
+            const deliCheckpoint: IDeliState = { ...checkpoint.deliState };
```
Comment (tanviraumi) [inline server/routerlicious/packages/lambdas/src/deli/checkpointContext.ts:89]: Just to double check, this does not change the checkpoint schema in any shape or form right? We still checkpoint the same thing that we used to so backward and forward compatibility should not get impacted?

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts [hunk 2]
```diff
@@ -141,15 +142,27 @@ export class DeliLambda implements IPartitionLambda {
...
         this.checkpointContext = new CheckpointContext(this.tenantId, this.documentId, checkpointManager, context);
+
+        // start the idle timer when created
+        this.setIdleTimer();
     }
 
     public handler(rawMessage: IQueuedMessage) {
+        let kafkaCheckpointMessage: IQueuedMessage | undefined;
+
         // In cases where we are reprocessing messages we have already checkpointed exit early
```
Comment (tanviraumi) [inline server/routerlicious/packages/lambdas/src/deli/lambda.ts:157]: This block requires better comment now. We should explain why we need to check for noActiveClient

**Review by tanviraumi — APPROVED** [review]

File: server/routerlicious/packages/lambdas/src/deli/checkpointContext.ts (hunk 1 above)
Comment (GaryWilber) [reply server/routerlicious/packages/lambdas/src/deli/checkpointContext.ts:89]: Correct. Before it was including the IQueuedMessage object in the checkpoint under a `queuedMessage` property. That property does not exist in `IDeliState`, so it was an extra property / unused data

File: server/routerlicious/packages/lambdas/src/deli/lambda.ts (hunk 2 above)
Comment (GaryWilber) [reply server/routerlicious/packages/lambdas/src/deli/lambda.ts:157]: I added a `getKafkaCheckpointMessage` method with a comment explaining it

**Review by tanviraumi — APPROVED** [review]

**Pull request merged**